src/nvim/normal.c

```diff
@@ -904,164 +877,347 @@ normal_cmd (
+    // This may mark the window for redrawing.
+    if (hasAnyFolding(curwin) && !char_avail()) {
+      foldCheckClose();
+
+      if (fdo_flags & FDO_ALL) {
+        foldOpenCursor();
+      }
+    }
+
+    // Before redrawing, make sure w_topline is correct, and w_leftcol
+    // if lines don't wrap, and w_skipcol if lines wrap.
+    update_topline();
+    validate_cursor();
+
+    if (VIsual_active) {
+      update_curbuf(INVERTED);  // update inverted part
+    } else if (must_redraw) {
+      update_screen(0);
+    } else if (redraw_cmdline || clear_cmdline) {
+      showmode();
+    }
+
+    redraw_statuslines();
+
+    if (need_maketitle) {
+      maketitle();
+    }
+
+    // display message after redraw
+    if (keep_msg != NULL) {
+      // msg_attr_keep() will set keep_msg to NULL, must free the string here.
+      // Don't reset keep_msg, msg_attr_keep() uses it to check for duplicates.
+      char *p = (char *)keep_msg;
+      msg_attr((uint8_t *)p, keep_msg_attr);
+      xfree(p);
+    }
+
+    if (need_fileinfo) {  // show file info after redraw
+      fileinfo(false, true, false);
+      need_fileinfo = false;
+    }
+
+    emsg_on_display = false;  // can delete error message now
+    did_emsg = false;
+    msg_didany = false;  // reset lines_left in msg_start()
+    may_clear_sb_text();  // clear scroll-back text on next msg
+    showruler(false);
+
+    if (s->conceal_update_lines
+        && (s->conceal_old_cursor_line !=
+          s->conceal_new_cursor_line
+          || conceal_cursor_line(curwin)
+          || need_cursor_line_redraw)) {
+      if (s->conceal_old_cursor_line !=
+          s->conceal_new_cursor_line
+          && s->conceal_old_cursor_line <=
+          curbuf->b_ml.ml_line_count) {
+        update_single_line(curwin, s->conceal_old_cursor_line);
+      }
+
+      update_single_line(curwin, s->conceal_new_cursor_line);
+      curwin->w_valid &= ~VALID_CROW;
+    }
+
+    setcursor();
+
+    do_redraw = false;
+
+    // Now that we have drawn the first screen all the startup stuff
+    // has been done, close any file for startup messages.
+    if (time_fd != NULL) {
+      TIME_MSG("first screen update");
+      TIME_MSG("--- NVIM STARTED ---");
+      fclose(time_fd);
+      time_fd = NULL;
+    }
+  }
+
+  // May perform garbage collection when waiting for a character, but
+  // only at the very toplevel.  Otherwise we may be using a List or
+  // Dict internally somewhere.
+  // "may_garbage_collect" is reset in vgetc() which is invoked through
+  // do_exmode() and normal_cmd().
+  may_garbage_collect = s->toplevel;
+
+  // Update w_curswant if w_set_curswant has been set.
+  // Postponed until here to avoid computing w_virtcol too often.
+  update_curswant();
 
-  /* Save count before an operator for next time. */
-  opcount = ca.opcount;
+  if (exmode_active) {
+    if (s->noexmode) {
+      return false;
+    }
+    do_exmode(exmode_active == EXMODE_VIM);
```

Comment:
If `!s->cmdwin || cmdwin_result == 0` is `true` then after the call `do_exmode()`, `normal_cmd()` is now always called. That was not the case in the old code.